File: android/lib/src/main/java/org/elastos/sdk/elephantwallet/contact/internal/ContactMessage.java
package org.elastos.sdk.elephantwallet.contact.internal;

import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.annotations.SerializedName;

import org.elastos.sdk.elephantwallet.contact.Contact;
import org.elastos.tools.crosspl.CrossBase;
import org.elastos.tools.crosspl.annotation.CrossClass;
import org.elastos.tools.crosspl.annotation.CrossInterface;

import java.io.File;

@CrossClass
public class ContactMessage extends CrossBase {
    public enum Type {
        Empty(0x00000000),
        MsgText(0x00000001),
        MsgAudio(0x00000002),
        MsgTransfer(0x00000004),
        MsgImage(0x00000008),
        MsgFile(0x00000010);

        public static Type valueOf(int id) {
            Type[] values = Type.values();
            for(int idx = 0; idx < values.length; idx++) {
                if(values[idx].id == id) {
                    return values[idx];
                }
            }
            return null;
        }

        private Type(int id){
            this.id = id;
        }
        private int id;
    }

    public static class MsgData {
        @Override
        public String toString() {
            String str = new Gson().toJson(this);
            return str;
        }
    }

    public static class TextData extends MsgData {
        public TextData(String text) {
            this.text = text;
        }

        @SerializedName(JsonKey.Text)
        final public String text;
    }

    public static class FileData extends MsgData {
        public FileData(File file) {
            devId = UserInfo.GetCurrDevId();
            name = file.getName();
            size = file.length();
            md5 = Utils.getMD5Sum(file);
        }

        // fix json decode and encode different issue
        public static String ConvertId(String id) {
            FileData fileData = new Gson().fromJson(id, FileData.class);
            if (fileData == null) {
                Log.w(Contact.TAG, "FileData.ConvertId() 0 Failed to convert " + id);
            }

            return fileData.toString();
        }

        @SerializedName(JsonKey.DeviceId)
        final public String devId;
        @SerializedName(JsonKey.Name)
        final public String name;
        @SerializedName(JsonKey.Size)
        final public long size;
        @SerializedName(JsonKey.Md5)
        final public String md5;
    }

    public final Type type;
    public MsgData data;
    public final String cryptoAlgorithm;
    public long timestamp;

    public int syncMessageToNative() {
        byte[] msgData = data.toString().getBytes();
        int ret = syncMessageToNative(type.id, msgData, cryptoAlgorithm, timestamp);
        return ret;
    }

    public ContactMessage(String text, String cryptoAlgorithm) {
        this(Type.MsgText, new TextData(text), cryptoAlgorithm);
    }

    public ContactMessage(File file, String cryptoAlgorithm) {
        this(Type.MsgFile, new FileData(file), cryptoAlgorithm);
    }

    public ContactMessage(Type type, byte[] data, String cryptoAlgorithm) {
        this(type, (MsgData) null, cryptoAlgorithm);
        if(data != null) {
            this.data = new Gson().fromJson(new String(data), GetDataClass(type));
        }
    }

    private ContactMessage(Type type, MsgData data, String cryptoAlgorithm) {
        super(ContactMessage.class.getName(), 0);

        this.type = type;
        this.data = data;
        this.cryptoAlgorithm = cryptoAlgorithm;
        this.timestamp = System.currentTimeMillis();
    }

    static private Class<? extends MsgData> GetDataClass(Type type) {
        switch (type) {
        case MsgText:
            return TextData.class;
        case MsgFile:
            return FileData.class;
        default:
            return null;
        }
    }

    @CrossInterface
    private native int syncMessageToNative(int type, byte[] data, String cryptoAlgorithm, long timestamp);
} // class Factory

File: ios/sdk/swift/internal/ContactMessage.swift
import CrossPL

/* @CrossClass */
open class ContactMessage: CrossBase {
  public enum Kind: Int {
    case Empty       = 0x00000000
    case MsgText     = 0x00000001
    case MsgAudio    = 0x00000002
    case MsgTransfer = 0x00000004
    case MsgImage    = 0x00000008
    case MsgFile     = 0x00000010
  }
  
  public class MsgData: Codable {
    public func toString() -> String {
      let encode = try! JSONEncoder().encode(self)
      let val = String(data: encode, encoding: .utf8)!
      return val
    }
    public func toData() -> Data {
      return String.ToData(from: self.toString())!
    }
  }

  public class TextData: MsgData {
    public init(text: String) {
      self.text = text;
      super.init()
    }
    
    required init(from decoder: Decoder) throws {
      let container = try decoder.container(keyedBy: CodingKeys.self)
      text = try container.decode(String.self, forKey: .text)
      try super.init(from: decoder)
    }
    
    override public func encode(to encoder: Encoder) throws {
      var container = encoder.container(keyedBy: CodingKeys.self)
      try container.encode(text, forKey: .text)
      try super.encode(to: encoder)
    }
    
    public private(set) var text: String?
    
    enum CodingKeys : String, CodingKey {
      case text
      var rawValue: String {
        get {
          switch self {
            case .text: return JsonKey.Text
          }
        }
      }
    }
  }

  public class FileData: MsgData {
    public init(file: URL) {
      devId = UserInfo.GetCurrDevId()!
      name = file.lastPathComponent
      size = try? FileManager.default.attributesOfItem(atPath: file.path)[FileAttributeKey.size] as? Int64
      md5 = Utils.getMD5Sum(file: file)
      super.init()
    }
    
    required init(from decoder: Decoder) throws {
      let container = try decoder.container(keyedBy: CodingKeys.self)
      devId = try container.decode(String.self, forKey: .devId)
      name = try container.decode(String.self, forKey: .name)
      size = try container.decode(Int64.self, forKey: .size)
      md5 = try container.decode(String.self, forKey: .md5)
      try super.init(from: decoder)
    }
    
    override public func encode(to encoder: Encoder) throws {
      var container = encoder.container(keyedBy: CodingKeys.self)
      try container.encode(devId, forKey: .devId)
      try container.encode(name, forKey: .name)
      try container.encode(size, forKey: .size)
      try container.encode(md5, forKey: .md5)
      try super.encode(to: encoder)
    }
    
    // fix json decode and encode different issue
    public static func ConvertId(id: String) -> String? {
      guard let dataId = id.data(using: .utf8) else {
        print("FileData.ConvertId() 0 Failed to convert \(id)")
        return nil
      }
      
      guard let fileData = try? JSONDecoder().decode(FileData.self, from: dataId) else {
        print("FileData.ConvertId() 1 Failed to convert \(id)")
        return nil
      }
      
      return fileData.toString()
    }
    
    public private(set) var devId: String
    public private(set) var name: String?
    public private(set) var size: Int64?
    public private(set) var md5: String?
    
    enum CodingKeys : String, CodingKey {
      case devId
      case name
      case size
      case md5
      var rawValue: String {
        get {
          switch self {
            case .devId: return JsonKey.DeviceId
            case .name: return JsonKey.Name
            case .size: return JsonKey.Size
            case .md5: return JsonKey.Md5
          }
        }
      }
    }
  }

  
  public let type: Kind
  public private(set) var data: MsgData
  public let cryptoAlgorithm: String?
  public var timestamp: Int64
  
  public func syncMessageToNative() -> Int {
    let ret = syncMessageToNative(type: type.rawValue,
                                  data: data.toData(),
                                  cryptoAlgorithm: cryptoAlgorithm,
                                  timestamp: timestamp)
    return ret;
  }


  public init(type: Kind, data: MsgData, cryptoAlgorithm: String?) {
    self.type = type
    self.data = data
    self.cryptoAlgorithm = cryptoAlgorithm ?? ""
    self.timestamp = Int64(Date().timeIntervalSince1970 * 1000)
    
    super.init(className: String(describing: ContactMessage.self))
  }

  public convenience init(text: String, cryptoAlgorithm: String?) {
    self.init(type: Kind.MsgText,
              data: TextData(text: text),
              cryptoAlgorithm: cryptoAlgorithm);
  }

  public convenience init(file: URL, cryptoAlgorithm: String?) {
    self.init(type: Kind.MsgFile,
              data: FileData(file: file),
              cryptoAlgorithm: cryptoAlgorithm);
  }
  
  public convenience init(type: Kind, data: Data, cryptoAlgorithm: String?) {
    self.init(type: type, data: MsgData(), cryptoAlgorithm: cryptoAlgorithm)
//      if(data != nil) {
        self.data = try! JSONDecoder().decode(ContactMessage.GetDataClass(type: type)!,
                                              from: data)
//      }
  }
  
  static private func GetDataClass(type: Kind) -> MsgData.Type? {
      switch (type) {
      case .MsgText:
        return TextData.self
      case .MsgFile:
        return FileData.self
      default:
        return nil
      }
  }
  
  /* @CrossNativeInterface */
  private func syncMessageToNative(type: Int,
                                   data: Data, cryptoAlgorithm: String?,
                                   timestamp: Int64) -> Int {
    let ret = crosspl_Proxy_ContactMessage_syncMessageToNative(nativeHandle, Int32(type), data, cryptoAlgorithm, timestamp)
    return Int(ret)
  }
}
